src/Thread.tsx

```diff
@@ -0,0 +1,55 @@
+import {defined} from './Base';
+import {fetchMessageIdsAndLabels, fetchMessagesByIds} from './Gapi';
+import {Message} from './Message';
+
+export class Thread {
+  private _rawThread: gapi.client.gmail.Thread;
+  private _messages: Message[] | undefined;
+  private _firstAndLastMessages: Message[] | undefined;
+
+  constructor(thread: gapi.client.gmail.Thread) {
+    this._rawThread = thread;
+    this._messages = undefined;
+    this._firstAndLastMessages = undefined;
+  }
+
+  get id(): string {
+    return defined(this._rawThread.id);
+  }
+
+  get messages(): Message[] {
+    return defined(this._messages);
+  }
+
+  get firstAndLastMessages(): Message[] | undefined {
+    return this._firstAndLastMessages;
+  }
+
+  hasMessagesInInbox(): boolean {
+    for (const message of defined(this.messages)) {
```

Comment:
return defined(this.messages).has(x => x.labelIds.includes('INBOX'))

I've started also using optional chaining in some places. It handles undefined gracefully rather than asserting. Not sure where we should land on that trade-off though. I like the readability though. :)

return this.messages?.has(x => x.labelIds.includes('INBOX'))